File: cart-service/src/main/java/org/vtsukur/graphql/demo/cart/web/graphql/spqr/GraphQLController.java
package org.vtsukur.graphql.demo.cart.web.graphql.spqr;

import graphql.ExecutionInput;
import graphql.ExecutionResult;
import graphql.GraphQL;
import graphql.analysis.MaxQueryComplexityInstrumentation;
import graphql.analysis.MaxQueryDepthInstrumentation;
import graphql.execution.batched.BatchedExecutionStrategy;
import graphql.execution.instrumentation.ChainedInstrumentation;
import graphql.schema.GraphQLSchema;
import io.leangen.graphql.GraphQLSchemaGenerator;
import io.leangen.graphql.metadata.strategy.query.AnnotatedResolverBuilder;
import io.leangen.graphql.metadata.strategy.query.BeanResolverBuilder;
import io.leangen.graphql.metadata.strategy.query.PublicResolverBuilder;
import io.leangen.graphql.metadata.strategy.value.jackson.JacksonValueMapperFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.ResponseBody;
import org.springframework.web.bind.annotation.RestController;

import java.util.Arrays;
import java.util.Map;

@RestController
public class GraphQLController {

    private final GraphQL graphQL;

    @Autowired
    public GraphQLController(CartGraph cartGraph, ProductGraph productQuery) {
        GraphQLSchema schema = new GraphQLSchemaGenerator()
                .withResolverBuilders(
                        new BeanResolverBuilder("org.volodymyrtsukur.playground.graphql.demo.domain.core"),
                        //Resolve by annotations
                        new AnnotatedResolverBuilder(),
                        //Resolve public methods inside root package
                        new PublicResolverBuilder("org.volodymyrtsukur.playground.graphql.demo.domain.integration.graphql_spqr"))
                .withOperationsFromSingleton(cartGraph)
                .withOperationsFromSingleton(productQuery)
                .withValueMapperFactory(new JacksonValueMapperFactory())
                .generate();
        graphQL = GraphQL.newGraphQL(schema)
                .queryExecutionStrategy(new BatchedExecutionStrategy())
                .instrumentation(new ChainedInstrumentation(Arrays.asList(
                        new MaxQueryComplexityInstrumentation(200),
                        new MaxQueryDepthInstrumentation(20)
                )))
                .build();
    }

    @PostMapping(value = "/graphql", consumes = MediaType.APPLICATION_JSON_UTF8_VALUE, produces = MediaType.APPLICATION_JSON_UTF8_VALUE)
    @ResponseBody
    public ExecutionResult execute(@RequestBody Map<String, Object> request) {
        return graphQL.execute(ExecutionInput.newExecutionInput()
                .query((String) request.get("query"))
                .operationName((String) request.get("operationName"))
                .build());
    }

}

File: cart-service/src/main/kotlin/org/vtsukur/graphql/demo/cart/web/graphql/spqr/GraphQLController.kt
package org.vtsukur.graphql.demo.cart.web.graphql.spqr

import graphql.ExecutionInput
import graphql.ExecutionResult
import graphql.GraphQL
import graphql.analysis.MaxQueryComplexityInstrumentation
import graphql.analysis.MaxQueryDepthInstrumentation
import graphql.execution.AsyncExecutionStrategy
import graphql.execution.instrumentation.ChainedInstrumentation
import graphql.execution.instrumentation.Instrumentation
import io.leangen.graphql.GraphQLSchemaGenerator
import io.leangen.graphql.metadata.strategy.query.AnnotatedResolverBuilder
import io.leangen.graphql.metadata.strategy.query.BeanResolverBuilder
import io.leangen.graphql.metadata.strategy.query.PublicResolverBuilder
import io.leangen.graphql.metadata.strategy.value.jackson.JacksonValueMapperFactory
import org.springframework.http.MediaType.APPLICATION_JSON_UTF8_VALUE
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.ResponseBody
import org.springframework.web.bind.annotation.RestController
import org.vtsukur.graphql.demo.cart.web.graphql.spqr.CartGraph
import org.vtsukur.graphql.demo.cart.web.graphql.spqr.ProductGraph
import java.util.*

@RestController
class GraphQLController (cartGraph: CartGraph, productQuery: ProductGraph) {

    private val graphQL: GraphQL

    init {
        val schema = GraphQLSchemaGenerator()
                .withResolverBuilders(
                        BeanResolverBuilder("org.vtsukur.graphql.demo.cart.domain"),
                        // Resolve by annotations.
                        AnnotatedResolverBuilder(),
                        // Resolve public methods inside root package.
                        PublicResolverBuilder("org.vtsukur.graphql.demo.cart.web.graphql.spqr"))
                .withOperationsFromSingleton(cartGraph)
                .withOperationsFromSingleton(productQuery)
                .withValueMapperFactory(JacksonValueMapperFactory())
                .generate()
        graphQL = GraphQL.newGraphQL(schema)
                .queryExecutionStrategy(AsyncExecutionStrategy())
                .instrumentation(ChainedInstrumentation(Arrays.asList<Instrumentation>(
                        MaxQueryComplexityInstrumentation(200),
                        MaxQueryDepthInstrumentation(20)
                )))
                .build()
    }

    @PostMapping(value = ["/graphql"], consumes = [APPLICATION_JSON_UTF8_VALUE], produces = [APPLICATION_JSON_UTF8_VALUE])
    @ResponseBody
    fun execute(@RequestBody request: Map<String, Any>): ExecutionResult =
            graphQL.execute(ExecutionInput.newExecutionInput()
                .query(request["query"] as String)
                .operationName((request["operationName"]?: "") as String)
                .build())

}
